### PR TITLE
Working prototype

### LINKED ISSUE #5: Implement a fully functional backend prototype
[Task board](https://github.com/demyanenko/hutel/projects/1)

Target features:
* Be cross-platform
* Accept new points and validate them against tags
* Store points on disk
* Store read-only tag config on disk
* Support string, int, float, date, time, enum fields in points

### PR DESCRIPTION
**Pull request by demyanenko**
This closes #5.

Changes:
* Points are persisted on disk
* Tags config is loaded from disk
* Custom type system supporting int, float, string, date, time and dynamic enums
* Reading/writing all points
* Filtering by date when getting all points

### CODE REVIEW

File: server/Models/Tag.cs [hunk 1]
```diff
@@ -13,8 +18,168 @@ public class Field
...
+        {
+            var jObject = JObject.Load(reader);
+            var id = jObject
+                .GetValue("id", StringComparison.OrdinalIgnoreCase)
+                .Value<string>();
+            var fieldsJson = (JArray)jObject.GetValue("fields", StringComparison.OrdinalIgnoreCase);
+            var fieldsList = fieldsJson.Select(fieldJson => DeserializeField((JObject)fieldJson));
+            foreach (var field in fieldsList)
+            {
+                if (Point.ReservedFields.Any(name => name == field.Name))
```
Comment (demyanenko) [inline server/Models/Tag.cs]: Comparison should ignore case

File: server/Controllers/PointsController.cs [hunk 2]
```diff
@@ -60,56 +110,45 @@ public IActionResult Post(Guid id, [FromBody]Point point)
                 return new BadRequestObjectResult(ex.ToString());
             }
             points[id] = point;
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: Please either write warning to console/logs or disallow to overwrite same point GUID

File: server/Logic/PointValidator.cs [hunk 3]
```diff
@@ -29,12 +29,9 @@ public static void Validate(Point point, Tag tag)
                 {
                     throw new ValidationException($"Property not found: {tagField.Name}");
                 }
-                var pointFieldType = point.Extra[tagField.Name].GetType();
-                if (pointFieldType != tagField.Type)
+                if (!tagField.Type.TypeIsValid(point.Extra[tagField.Name]))
```
Comment (Emettant) [inline server/Logic/PointValidator.cs]: this one check fail to validate type of incoming time and date type, please see comments below in tags.json.example

File: server/tags.json.example [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    },
+    {
+        "id": "vacation",
+        "fields": [
+            {
+                "name": "location",
+                "type": "string"
+            },
+            {
+                "name": "enddate",
```
Comment (demyanenko) [inline server/tags.json.example]: Correct date format has all zeros in time part. I'll write down the api specs.

File: server/tags.json.example [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        "id": "run",
+        "fields": [
+            {
+                "name": "km",
+                "type": "float"
+            },
+            {
+                "name": "time",
+                "type": "time"
```
Comment (demyanenko) [inline server/tags.json.example]: There was a fault in type validation logic for time fields. Now the correct time is represented as ISO 8601 date string with date set to 0001-01-01.

File: server/Logic/PointValidator.cs (hunk 3 above)
Comment (demyanenko) [inline server/Logic/PointValidator.cs]: Date logic was correct (but unobvious), time validation fixed

**Comment by demyanenko** [comment]
To do: refactor until the code is testable, write unit tests for validation

File: server/Controllers/PointsController.cs [hunk 6]
```diff
@@ -109,6 +115,10 @@ public IActionResult PostOne(Guid id, [FromBody]Point point)
             {
                 return new BadRequestObjectResult(ex.ToString());
             }
+            if (points.ContainsKey(point.Id))
+            {
+                _logger.LogWarning($"Overwriting the point with id {point.Id}");
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: nit: is this the only place we need logger; what do you think to use it for some exceptions (if not for all exceptions)?

File: server/Controllers/PointsController.cs [hunk 7]
```diff
@@ -6,21 +6,26 @@
...
-        private IMemoryCache _memoryCache;
+        private readonly IMemoryCache _memoryCache;
+        private readonly ILogger _logger;
         private const string _pointsKey = "points";
         private const string _tagsKey = "tags";
-        private const string _storagePath = ".\\server\\storage.json";
-        private const string _storageBackupPath = ".\\server\\storage.json.bak";
-        private const string _tagsPath = ".\\server\\tags.json";
+        private static readonly string _storagePath =
+            System.IO.Path.Combine(".", "server", "storage.json");
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: I do:
 cd ./server
dotnet build
dotnet run

in this configuration my relative pathes are  System.IO.Path.Combine(".", "storage.json") ; System.IO.Path.Combine(".", "storage.json.bak");  System.IO.Path.Combine(".", "tags.json");

So I mean they must not have "server" part in the relative paths.

File: server/Models/Tag.cs [hunk 8]
```diff
@@ -62,10 +62,11 @@ public bool TypeIsValid(object obj)
         
         public class FTime : IFieldType
         {
+            private static readonly DateTime _epoch = new DateTime(1, 1, 1);
```
Comment (Emettant) [inline server/Models/Tag.cs]: nit: I have your comment on prev commit about code convention here, however without reading that there is no docs describing this convention (comments or mentioning in readme file might be good ideas)

File: server/Controllers/PointsController.cs [hunk 9]
```diff
@@ -45,6 +50,7 @@ public PointsController(IMemoryCache memoryCache)
                         .SetPriority(CacheItemPriority.NeverRemove)
                 );
             }
+            _logger = logger;
         }
 
         // GET /api/points
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: what do we do for next situation:
 1. tags.json contains tag "run"
 2. one point for "run" has been submitted, validated and stored to storage.json
 3. tags.json has been modified manually: "run" was replaced by "walk"
 4. one point for "walk" has been submitted, validated and stored to storage.json
 5. I GET all points from storage.json to a user buffer.
 6. I do PostAll same points from the user buffer
 7. I get validation exceptions.

**Comment by demyanenko** [comment]
Fixed all the issues. Please let me know if anything has been missed. Also, what kind of documentation should we include? The web-facing API only or something else as well?

File: server/Controllers/PointsController.cs [hunk 10]
```diff
@@ -117,26 +110,30 @@ public IActionResult PostOne(Guid id, [FromBody]Point point)
                 _logger.LogWarning($"Overwriting the point with id {point.Id}");
             }
             points[id] = point;
-            WriteStorage(points);
-            return Json(points.Values);
+            WriteStorage(points, tags);
+            return Json(points.Values.Select(p => p.ToJson(tags)).ToList());
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: returns all the values, however it seems it should be the only one :) has been talked offline

File: server/Logic/HutelDate.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public class HutelDate
+    {
+        private DateTime DateTime { get; }
+
+        private const string _format = "yyyy-MM-dd";
+
+        public HutelDate(string date)
+        {
+            DateTime = DateTime.ParseExact(date, _format, CultureInfo.InvariantCulture);
```
Comment (Emettant) [inline server/Logic/HutelDate.cs]: try:  "date": "2017-09-31"
dont throw even general exception,
so BadRequest answer is not formed.
I am not sure if this is expected behaviour

File: server/Logic/HutelTime.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public class HutelTime
+    {
+        public TimeSpan TimeSpan { get; }
+
+        private const string _format = @"hh\:mm\:ss";
+
+        public HutelTime(string time)
+        {
+            TimeSpan = TimeSpan.ParseExact(time, _format, CultureInfo.InvariantCulture);
```
Comment (Emettant) [inline server/Logic/HutelTime.cs]: "time": "24:55:20",
"time": "23:60:20",

File: server/Controllers/PointsController.cs [hunk 13]
```diff
@@ -52,63 +52,56 @@ public PointsController(IMemoryCache memoryCache, ILogger<PointsController> logg
 
         // GET /api/points
         [HttpGet]
-        public IActionResult GetAll(DateTime startDate)
+        public IActionResult GetAll(string startDate)
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: Same 500 Internal Server Error
http://localhost:5000/api/points?startDate=2017-09-31

File: server/Controllers/PointsController.cs [hunk 14]
```diff
@@ -52,63 +52,56 @@ public PointsController(IMemoryCache memoryCache, ILogger<PointsController> logg
...
-            catch(ValidationException ex)
+            catch(PointValidationException ex)
             {
                 return new BadRequestObjectResult(ex.ToString());
             }
             points.Clear();
-            foreach (var p in replacementPoints)
+            foreach (var p in pointsList)
             {
                 points.Add(p.Id, p);
```
Comment (Emettant) [inline server/Controllers/PointsController.cs]: Same ids cand be added to Dictionary
[
{
    "id": "044af706-92e7-435c-b788-cd8370938a11",
   ...
  },
  {
    "id": "044af706-92e7-435c-b788-cd8370938a11",
...
  }
]

File: server.test/FieldTypeTests.cs [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            get
+            {
+                return new[]
+                {
+                    new Object[] { typeof(IntFieldType), "" },
+                    new Object[] { typeof(FloatFieldType), "" },
+                    new Object[] { typeof(StringFieldType), 0 },
+                    new Object[] { typeof(DateFieldType), "01/02/2000" },
+                    new Object[] { typeof(DateFieldType), "01012000" },
```
Comment (Emettant) [inline server.test/FieldTypeTests.cs]: new Object[] { typeof(DateFieldType), "2017-09-31" },
new Object[] { typeof(TimeFieldType), "24:00:00" },
new Object[] { typeof(TimeFieldType), "00:60:00" }

**Pull request merged**